src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,227 @@
+                        void ReportIfTypesAreDifferent(IParameterSymbol parameter, ITypeSymbol suggestedBaseTypeSymbol)
+                        {
+                            if (!SymbolEqualityComparer.Default.Equals(suggestedBaseTypeSymbol, parameter.Type))
+                            {
+                                // All calls on this parameter are using the same base type which is different
+                                // from the one declared, so we can simply report.
+                                context.ReportDiagnostic(parameter.CreateDiagnostic(Rule, parameter.Name,
+                                    parameter.Type.Name, suggestedBaseTypeSymbol.Name));
+                            }
+                        }
+                    });
+                });
+            });
+        }
+
+        private static ITypeSymbol FindDeclaringType(ISymbol symbol)
+        {
+            Debug.Assert(
+                symbol.Kind != SymbolKind.Method ||
+                symbol.Kind != SymbolKind.Property ||
+                symbol.Kind != SymbolKind.Event);
+
+            if (symbol.IsOverride)
+            {
+                return FindDeclaringType(symbol.GetOverriddenMember());
+            }
+
+            var interfaceMembers = ExplicitOrImplicitInterfaceImplementations(symbol);
+            if (interfaceMembers.Length == 1)
+            {
+                return interfaceMembers[0].ContainingType;
+            }
+            else
+            {
+                // When reaching this point, either we found no interface member, which means the
+                // member is declared directly on the type OR more than one interface was found
+                // as declaring this type so let's assume it is defined on the containing type.
+                return symbol.ContainingType;
+            }
+        }
+
+        // Borrowed from Microsoft.CodeAnalysis.Workspaces/ISymbolExtensions.cs with a small tweak
+        // which consist in passing the member name.
+        private static ImmutableArray<ISymbol> ExplicitOrImplicitInterfaceImplementations(ISymbol symbol)
+        {
+            if (symbol.Kind != SymbolKind.Method &&
+                symbol.Kind != SymbolKind.Property &&
+                symbol.Kind != SymbolKind.Event)
+            {
+                return ImmutableArray<ISymbol>.Empty;
+            }
+
+            var containingType = symbol.ContainingType;
+            var query = from iface in containingType.AllInterfaces
+                        from interfaceMember in iface.GetMembers(symbol.Name)
+                        let impl = containingType.FindImplementationForInterfaceMember(interfaceMember)
+                        where symbol.Equals(impl)
+                        select interfaceMember;
+            return query.ToImmutableArray();
+        }
+
+        private static ITypeSymbol FindMostGenericType(ITypeSymbol originalType, ImmutableArray<ITypeSymbol> constraints)
```

Comment:
The ordering in this method always favors interfaces over types, this could be revisited if wanted.